Cilium Network Policy v3

aanm opened the pull request:
**Summary of changes**:

Policy v3 implementation

Fixes: #3005 

```release-note
New Cilium Network Policy v3
```

There's lack of documentation since it's WIP. Please suggest better fields names for the Network Policy attributes if needed.

There's a translator from v2 to v3. It's only missing some unit tests.

```yaml
---
version: v3
rules:
- identitySelector:
    matchLabels:
      any:id: backend
  ingress:
  - fromIdentities:
      identitySelector:
        matchLabels:
          any:id: frontend
        matchExpressions:
        - key: any:id
          operator: In
          values:
          - frontend
          - qa
  - fromIdentities:
      identitySelector:
        matchLabels:
          any:id: team-a
      toPorts:
        anyOf:
        - port: '111'
          protocol: TCP
  - fromCIDR:
      anyOf:
      - 1.1.1.1/31
      except:
      - 1.1.1.1/32
      toPorts:
        anyOf:
        - port: '9090'
          protocol: TCP
  egress:
  - toServices:
      k8sServiceSelector:
        serviceSelector:
          matchLabels:
            id: backup
  - toIdentities:
      identitySelector:
        matchLabels:
          any:group: backend
  - toIdentities:
      identitySelector:
        matchLabels:
          any:id: qa
  - toRequires:
      anyOf:
      - matchLabels:
          any:team: b
      - matchLabels:
          any:team: c
  - toCIDR:
      anyOf:
      - 1.1.1.1/31
      except:
      - 1.1.1.1/32
      toPorts:
        anyOf:
        - port: '9191'
          protocol: TCP
        rules:
          http:
          - path: "/"
            method: GET
            headers:
            - foo=bar
  - toEntities:
      anyOf:
      - world
      toPorts:
        anyOf:
        - port: '8080'
          protocol: TCP
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/3164)
<!-- Reviewable:end -->

aanm committed:
refactored policy/api to policy/api/v2

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/api/v3: introduced policy v3

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/k8s: refactor const.go to common directory

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/k8s: add k8s cnp/v3 object
k8s: update rbac rules for v3 policy

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,244 @@
+	// Labels is a list of optional strings which can be used to
+	// re-identify the rule or to store metadata. It is possible to lookup
+	// or delete strings based on labels. Labels are not required to be
+	// unique, multiple rules can have overlapping or identical labels.
```
unique;

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,244 @@
+	// +optional
+	Labels labels.LabelArray `json:"labels,omitempty"`
+
+	// Description is a free form string, it can be used by the creator of
```
free-form string. It can be used by...

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,244 @@
+	Labels labels.LabelArray `json:"labels,omitempty"`
+
+	// Description is a free form string, it can be used by the creator of
+	// the rule to store human readable explanation of the purpose of this
```
... to store a human-readable explanation...

aanm committed:
fix test failures for the v3 policy

Ignored some test failures and opened relevanted GH issues.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
convert all cilium code to use v3 policy internally

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on cilium/cmd/policy.go:
```diff
@@ -26,7 +26,7 @@ import (
 
 	"github.com/cilium/cilium/pkg/logging/logfields"
-	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/pkg/policy/api/v2"
```
Title of first commit has a typo

tgraf commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+	ToPorts          *PortRule        `json:"toPorts,omitempty"`
+}
+
+type EndpointRequire struct {
```
Rename to `IdentityRequirement`

tgraf commented on pkg/policy/api/v3/selector.go:
```diff
@@ -0,0 +1,215 @@
+var log = logging.DefaultLogger
+
+// IdentitySelector is a wrapper for k8s LabelSelector.
+type IdentitySelector struct {
```
Change to `LabelSelector` if possible without major surgery

tgraf commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,507 @@
+
+	v3Rule.EndpointSelector = *v2ESTov3ES(&v2Rule.EndpointSelector)
+
+	if v3Rule.Ingress != nil {
```
1. v2
2. move for loop into branch

tgraf commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,507 @@
+		v3Rule.Ingress = append(v3Rule.Ingress, v2IRTov3IR(&v)...)
+	}
+
+	if v3Rule.Egress != nil {
```
1. v2
2. move for loop into branch

tgraf commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,507 @@
+	}
+
+	v3Rule.Labels = v2Rule.Labels.DeepCopy()
+
```
kill line

tgraf commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,507 @@
+	if v2IR.ToPorts != nil {
+		v3PR = make([]*PortRule, len(v2IR.ToPorts))
+	}
+	for i, v := range v2IR.ToPorts {
```
Move into branch

tgraf commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,507 @@
+	"github.com/cilium/cilium/pkg/policy/api/v2"
+)
+
+func V2RulesTov3Rules(v2Rules *v2.Rules) *Rules {
```
Add comment that v2 must have been sanitized

tgraf commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,507 @@
+				IngressRule{FromIdentities: &IdentityRule{IdentitySelector: NewWildcardIdentitySelector(), ToPorts: v.DeepCopy()}},
+			)
+		}
+
```
empty line

tgraf commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,507 @@
+				EgressRule{ToIdentities: &IdentityRule{IdentitySelector: NewWildcardIdentitySelector(), ToPorts: v.DeepCopy()}},
+			)
+		}
+
```
empty line

tgraf commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,507 @@
+	if v2CR.ExceptCIDRs != nil {
+		v3CR.ExceptCIDRs = make([]CIDR, len(v2CR.ExceptCIDRs))
+	}
+	for i, v := range v2CR.ExceptCIDRs {
```
move into branch

tgraf commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,507 @@
+		APIVersion:    v2K.APIVersion,
+		ClientID:      v2K.ClientID,
+		Topic:         v2K.Topic,
+		apiKeyInt:     v3KafkaRole,
```
Remove translation of private fields

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -1404,11 +1457,12 @@ func (d *Daemon) addCiliumNetworkPolicyV2(ciliumV2Store cache.Store, cnp *cilium
 
 	rules, err := cnp.Parse()
 	if err == nil && len(rules) > 0 {
+		v3Rules := v3.V2RulesTov3Rules(&rules)
```
Needs to sanitise the rule again after convertion

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -1506,14 +1560,15 @@ func (d *Daemon) deleteCiliumNetworkPolicyV2(cnp *cilium_v2.CiliumNetworkPolicy)
 	}
 
 	rules, err := cnp.Parse()
+	v3Rules := v3.V2RulesTov3Rules(&rules)
```
Sanitise rules again

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -1506,14 +1560,15 @@ func (d *Daemon) deleteCiliumNetworkPolicyV2(cnp *cilium_v2.CiliumNetworkPolicy)
 	}
 
 	rules, err := cnp.Parse()
```
Check error before translation

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -1556,6 +1611,171 @@ func (d *Daemon) updateCiliumNetworkPolicyV2(ciliumV2Store cache.Store,
+					return fmt.Errorf("unable to find v2.CiliumNetworkPolicy in local cache: %s", err2)
+				}
+				if !exists {
+					return errors.New("v2.CiliumNetworkPolicy does not exist in local cache")
```
v2 -> v3

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -1556,6 +1611,171 @@ func (d *Daemon) updateCiliumNetworkPolicyV2(ciliumV2Store cache.Store,
+
+				serverRuleStore, exists, err2 := ciliumV2Store.Get(cnp)
+				if err2 != nil {
+					return fmt.Errorf("unable to find v2.CiliumNetworkPolicy in local cache: %s", err2)
```
v2 -> v3

tgraf commented on pkg/k8s/network_policy.go:
```diff
@@ -107,72 +107,133 @@ func hasV1PolicyType(pTypes []networkingv1.PolicyType, typ networkingv1.PolicyTy
+					if len(l4Ports) == 0 {
+						ingresses = append(ingresses, v3.IngressRule{
+							FromIdentities: &v3.IdentityRule{
+								IdentitySelector: *identitySelector,
```
Add comment to add L4 wildcard here (Thomas)

tgraf commented on pkg/k8s/network_policy.go:
```diff
@@ -107,72 +107,133 @@ func hasV1PolicyType(pTypes []networkingv1.PolicyType, typ networkingv1.PolicyTy
-					ingress.FromCIDRSet = append(ingress.FromCIDRSet, ipBlockToCIDRRule(rule.IPBlock))
+					if len(l4Ports) == 0 {
+						ingresses = append(ingresses, v3.IngressRule{
+							FromCIDRs: ipBlockToCIDRRule(rule.IPBlock),
```
Add comment to add L4 wildcard here (Thomas)

tgraf commented on pkg/k8s/network_policy.go:
```diff
@@ -107,72 +107,133 @@ func hasV1PolicyType(pTypes []networkingv1.PolicyType, typ networkingv1.PolicyTy
-			ingress.FromEndpoints = append(ingress.FromEndpoints, all)
+			ingresses = append(ingresses, v3.IngressRule{
+				FromIdentities: &v3.IdentityRule{
+					IdentitySelector: all,
```
Add comment to add L4 wildcard here (Thomas)

tgraf commented on pkg/k8s/network_policy.go:
```diff
@@ -107,72 +107,133 @@ func hasV1PolicyType(pTypes []networkingv1.PolicyType, typ networkingv1.PolicyTy
 	for _, iRule := range np.Spec.Ingress {
-		ingress := v2.IngressRule{}
+
+		l4Ports := parsePorts(iRule.Ports)
```
This can return `nil` if empty list or a `DeepCopy()` so we can call it from all places that need to fill `ToPorts`

tgraf commented on pkg/k8s/network_policy.go:
```diff
@@ -107,72 +107,133 @@ func hasV1PolicyType(pTypes []networkingv1.PolicyType, typ networkingv1.PolicyTy
-		if iRule.Ports != nil && len(iRule.Ports) > 0 {
-			ingress.ToPorts = parsePorts(iRule.Ports)
-		} else if iRule.From == nil || len(iRule.From) == 0 {
+		if len(iRule.Ports) == 0 && len(iRule.From) == 0 {
```
Thomas will fix logic in v2, Andre to readjust after merging v2 fix

tgraf commented on pkg/k8s/network_policy.go:
```diff
@@ -185,7 +246,12 @@ func ParseNetworkPolicy(np *networkingv1.NetworkPolicy) (v2.Rules, error) {
 			!hasV1PolicyType(np.Spec.PolicyTypes, networkingv1.PolicyTypeEgress)) {
-		ingresses = []v2.IngressRule{{}}
+
+		ingresses = []v3.IngressRule{
```
No comment necessary

tgraf commented on pkg/k8s/network_policy.go:
```diff
@@ -194,16 +260,21 @@ func ParseNetworkPolicy(np *networkingv1.NetworkPolicy) (v2.Rules, error) {
 	if len(egresses) == 0 && hasV1PolicyType(np.Spec.PolicyTypes, networkingv1.PolicyTypeEgress) {
-		egresses = []v2.EgressRule{{}}
+
+		egresses = []v3.EgressRule{
```
No comment necessary again

tgraf commented on pkg/policy/rule.go:
```diff
@@ -253,12 +255,15 @@ func (r *rule) resolveL4Policy(ctx *SearchContext, state *traceState, result *L4
-			if cnt > 0 {
-				found += cnt
+			if ingressRule.FromIdentities != nil {
+				cnt, err := mergeL4(ctx, "Ingress", &ingressRule.FromIdentities.IdentitySelector,
```
Consider not passing a pointer in

tgraf commented on pkg/policy/rule.go:
```diff
@@ -268,12 +273,16 @@ func (r *rule) resolveL4Policy(ctx *SearchContext, state *traceState, result *L4
-			}
-			if cnt > 0 {
-				found += cnt
+			// FIXME FIXME FIXME why no toEndpoints Selector?
```
Known, Ian is working on it

tgraf commented on pkg/policy/rule.go:
```diff
@@ -164,66 +168,64 @@ func mergeL4Port(ctx *SearchContext, fromEndpoints []v2.EndpointSelector, r v2.P
-				ctx.PolicyTrace("        %+v\n", l7)
-			}
+	l3match := false
+	if ctx.From != nil && fromEndpoints != nil {
```
remove `fromEndpoints != nil`

tgraf commented on pkg/policy/rule.go:
```diff
@@ -164,66 +168,64 @@ func mergeL4Port(ctx *SearchContext, fromEndpoints []v2.EndpointSelector, r v2.P
-			}
+	l3match := false
+	if ctx.From != nil && fromEndpoints != nil {
+		if fromEndpoints.Matches(ctx.From) {
```
```
if !fromEndpoints.Matches(ctx.From) {
   ctx.PolicyTrace("      Labels %s not found", ctx.From)
   return 0, nil
}
```

tgraf commented on pkg/policy/rule.go:
```diff
@@ -385,37 +396,42 @@ func (r *rule) resolveCIDRPolicy(ctx *SearchContext, state *traceState, result *
-					ctx.PolicyTrace("+       No L4 restrictions\n")
-					state.matchedRules++
-					return v2.Allowed
+				if r.FromIdentities.ToPorts != nil {
```
Move this logic into a `IsWildcard()` function which returns true if `ToPorts` is nil or `ToPorts` slice is empty.

Add FIXME that this doe snot respect protocol yet

tgraf commented on pkg/policy/rule.go:
```diff
@@ -428,76 +444,79 @@ func (r *rule) canReachIngress(ctx *SearchContext, state *traceState) v2.Decisio
+			return v3.Allowed
 		}
 
 	}
```
Add FIXME here to respect CIDR in the future

tgraf replied:
Sorry, we need to remove this again.

tgraf commented on pkg/policy/rule.go:
```diff
@@ -428,76 +444,79 @@ func (r *rule) canReachIngress(ctx *SearchContext, state *traceState) v2.Decisio
 		if entitySelector.Matches(ctx.From) {
 			ctx.PolicyTrace("+     Found all required labels to match entity %s\n", entitySelector.String())
 			state.matchedRules++
```
Add the L4 restriction code from identities above  here as well

aanm replied:
I can't, the `fromEntities` is not coming from the `v3.Rule`, is coming from the `policy.Rule`. I added a FIXME

tgraf commented on pkg/policy/rule.go:
```diff
@@ -428,76 +444,79 @@ func (r *rule) canReachIngress(ctx *SearchContext, state *traceState) v2.Decisio
+		ctx.PolicyTrace("    Allows to labels %+v", sel)
+		if sel.Matches(ctx.To) {
+			ctx.PolicyTrace("      Found all required labels")
+			if r.ToIdentities.ToPorts != nil {
```
Same here

tgraf commented on pkg/policy/rule.go:
```diff
@@ -428,76 +444,79 @@ func (r *rule) canReachIngress(ctx *SearchContext, state *traceState) v2.Decisio
 	for _, entitySelector := range r.toEntities {
 		if entitySelector.Matches(ctx.To) {
 			ctx.PolicyTrace("+     Found all required labels to match entity %s\n", entitySelector.String())
 			state.matchedRules++
```
Same here

tgraf reviewed (changes requested):
Given we are moving to cnp.cilium.io, we could start counting at v1 if we wanted to. I would be a cleaner but probably requires major surgery in the PR.

nebril commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,240 @@
+
+	ToEntities *EntityRule `json:"toEntities,omitempty"`
+
+	ToServices *ServiceRule `json:"toServices,omitempty"`
```
This is no longer a slice, should this be renamed ToService? It can still match multiple services via selector, so I am not sure about it.

nebril commented on pkg/policy/api/v3/rule_validation.go:
```diff
@@ -0,0 +1,327 @@
+}
+
+func (e *EgressRule) sanitize() error {
+	l3Members := map[string]bool{
```
Should e.ToRequires also be sanitized here?

nebril commented on pkg/policy/api/v3/rule_validation.go:
```diff
@@ -0,0 +1,327 @@
+		return 0, fmt.Errorf("IP must be specified")
+	}
+
+	//_, ipnet, err := net.ParseCIDR(strCIDR)
```
Please remove commented out code

nebril commented on pkg/policy/api/v3/rules.go:
```diff
@@ -0,0 +1,22 @@
+// All rules must be evaluated in order to come to a conclusion. While
+// it is sufficient to have a single fromEndpoints rule match, none of
+// the fromRequires may be violated at the same time.
+type Rules []*Rule
```
Does this type require separate file? I am OK with that, just asking what's the reason for that.

nebril commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,539 @@
+
+// V2RuleTov3RuleSanitized translates the given v2Rule to the respective v3
+// Rule.
+// The given v2Rules needs to be sanitized and the returned v3 Rule is alread
```
s/alread/already

nebril reviewed (commented):
ToServices bits look OK, one rename suggestion and several random nitpicks.

tgraf commented on pkg/k8s/network_policy.go:
```diff
@@ -127,47 +127,31 @@ func ParseNetworkPolicy(np *networkingv1.NetworkPolicy) (v3.Rules, error) {
 			}
 		}
-		if len(iRule.Ports) == 0 && len(iRule.From) == 0 {
+		if len(iRule.From) == 0 {
```
Could do else branch here

tgraf commented on pkg/k8s/network_policy.go:
```diff
@@ -193,42 +178,25 @@ func ParseNetworkPolicy(np *networkingv1.NetworkPolicy) (v3.Rules, error) {
+					egresses = append(egresses, toCIDRs)
 				}
 			}
 		}
```
Could do else branch with below branch

tgraf commented on pkg/policy/repository_test.go:
```diff
@@ -354,33 +358,38 @@ func buildSearchCtx(from, to string, port uint16) *SearchContext {
+	expectedVerdict v3.Decision) {
 
 	buffer := new(bytes.Buffer)
 	ctx.Logging = logging.NewLogBackend(buffer, "", 0)
```
Add a `c.Log(buffer)` here so we have the trace output

tgraf commented on pkg/policy/repository_test.go:
```diff
@@ -390,16 +399,17 @@ func (repo *Repository) checkTrace(c *C, ctx *SearchContext, trace string,
 
-	expectedOut := "Tracing " + ctx.String() + trace
-	c.Assert(buffer.String(), comparator.DeepEquals, expectedOut)
+	// FIXME FIXME FIXME FIXME FIXME FIXME FIXME
```
After that, remove this.

tgraf commented on pkg/policy/rule.go:
```diff
@@ -313,7 +305,14 @@ func computeResultantCIDRSet(cidrs *v3.CIDRRule) []v3.CIDR {
 	var allResultantAllowedCIDRs []v3.CIDR
 	for _, c := range cidrs.CIDR {
 		// No need for error checking, as v3.CIDRRule.Sanitize() already does.
```
Remove this comment

tgraf commented on pkg/policy/rule.go:
```diff
@@ -313,7 +305,14 @@ func computeResultantCIDRSet(cidrs *v3.CIDRRule) []v3.CIDR {
-		_, allowNet, _ := net.ParseCIDR(string(c))
+		_, allowNet, err := net.ParseCIDR(string(c))
+		if err != nil {
+			ip := net.ParseIP(string(c))
```
Add comment here that we are trying to parse `"x"` instead of `"x/y"`

tgraf commented on pkg/policy/rule.go:
```diff
@@ -442,13 +444,15 @@ func (r *rule) canReachIngress(ctx *SearchContext, state *traceState) v3.Decisio
 
 	for _, entitySelector := range r.fromEntities {
```
This should be using `FromEntities` of the ingress rule.

tgraf replied:
Sanitisation should happen in `api/v3`

tgraf commented on pkg/policy/rule.go:
```diff
@@ -481,31 +485,35 @@ func (r *rule) canReachEgress(ctx *SearchContext, state *traceState) v3.Decision
 		}
 	}
 
 	for _, entitySelector := range r.toEntities {
```
Same applies here

tgraf commented on pkg/policy/rule.go:
```diff
@@ -481,31 +485,35 @@ func (r *rule) canReachEgress(ctx *SearchContext, state *traceState) v3.Decision
 		}
 	}
 
+	// FIXME respect CIDR in the future
```
Remove again

tgraf commented on pkg/policy/rule_test.go:
```diff
@@ -219,22 +223,23 @@ func (ds *PolicyTestSuite) TestL4Policy(c *C) {
 	expected.Egress["3000/TCP"] = L4Filter{
-		Port: 3000, Protocol: v2.ProtoTCP, U8Proto: 6, Ingress: false,
-		L7RulesPerEp:     L7DataMap{},
+		Port: 3000, Protocol: v3.ProtoTCP, U8Proto: 6, FromEndpoints: []v3.IdentitySelector{},
```
Extend `IsWildcard()` to check for `nil`

ianvernon commented on pkg/logging/logfields/logfields.go:
```diff
@@ -123,8 +123,11 @@ const (
+	// IdentityLabelSelector is a selector for Identities by label
+	IdentityLabelSelector = "IdentityLabelSelector"
+
+	// IdentitySelector is a selector for Endpoints
```
Fix this naming? It is confusing.

ianvernon commented on pkg/policy/api/v3/decision.go:
```diff
@@ -0,0 +1,73 @@
+	"fmt"
+)
+
+// Decision is a reachability policy decision
```
... policy-reachability decision.

ianvernon commented on pkg/policy/api/v3/decision.go:
```diff
@@ -0,0 +1,73 @@
+type Decision byte
+
+const (
+	// Undecided means that we have not come to a decision yet
```
I don't like the use of 'we' in public APIs. 
"... specifies that whether policy allows or denies traffic is not able to be determined." ?

ianvernon commented on pkg/policy/api/v3/decision.go:
```diff
@@ -0,0 +1,73 @@
+	Undecided Decision = iota
+	// Allowed means that reachability is allowed
+	Allowed
+	// Denied means that reachability is denied
```
denied.

ianvernon commented on pkg/policy/api/v3/decision.go:
```diff
@@ -0,0 +1,73 @@
+const (
+	// Undecided means that we have not come to a decision yet
+	Undecided Decision = iota
+	// Allowed means that reachability is allowed
```
allowed.

ianvernon commented on pkg/policy/api/v3/decision.go:
```diff
@@ -0,0 +1,73 @@
+	}
+)
+
+// String returns the decision in human readable format
```
human-readable format.

ianvernon commented on pkg/policy/api/v3/decision.go:
```diff
@@ -0,0 +1,73 @@
+	return ""
+}
+
+// UnmarshalJSON parses a JSON formatted buffer and returns a decision
```
" ... attempts to unmarshal the contents of b into a Decision". 

Also, this does not return a decision, it returns an error.

ianvernon commented on pkg/policy/api/v3/decision.go:
```diff
@@ -0,0 +1,73 @@
+	return fmt.Errorf("unknown '%s' decision", string(b))
+}
+
+// MarshalJSON returns the decision as JSON formatted buffer
```
as a JSON-formatted buffer.

ianvernon commented on pkg/policy/api/v3/doc.go:
```diff
@@ -0,0 +1,20 @@
+// +k8s:deepcopy-gen=package
+// +k8s:openapi-gen=true
+
+// Package api defines the API of the Cilium network policy interface
```
Fix this documentation

ianvernon commented on pkg/policy/api/v3/http.go:
```diff
@@ -0,0 +1,58 @@
+package v3
+
+// PortRuleHTTP is a list of HTTP protocol constraints. All fields are
+// optional, if all fields are empty or missing, the rule does not have any
```
... optional; if all fields ...

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+)
+
+// PortRuleKafka is a list of Kafka protocol constraints. All fields are
+// optional, if all fields are empty or missing, the rule will match all
```
... optional; if all fields...

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+// Kafka messages.
+type PortRuleKafka struct {
+	// Role is a case-insensitive string and describes a group of API keys
+	// necessary to perform certain higher level Kafka operations such as "produce"
```
higher-level

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+	// Role is a case-insensitive string and describes a group of API keys
+	// necessary to perform certain higher level Kafka operations such as "produce"
+	// or "consume". An APIGroup automatically expands into all APIKeys required
+	// to perform the specified higher level operation.
```
higher-level

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+	//  - "produce": Allow producing to the topics specified in the rule
+	//  - "consume": Allow consuming from the topics specified in the rule
+	//
+	// This field is incompatible with the APIKey field, either APIKey or Role
```
... field; either ...

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+	Role string `json:"role,omitempty"`
+
+	// APIKey is a case-insensitive string matched against the key of a
+	// request, e.g. "produce", "fetch", "createtopic", "deletetopic", et al
```
et al --> et al.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+	// ClientID is the client identifier as provided in the request.
+	//
+	// From Kafka protocol documentation:
+	// This is a user supplied identifier for the client application. The
```
user-supplied

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+	// This is a user supplied identifier for the client application. The
+	// user can use any identifier they like and it will be used when
+	// logging errors, monitoring aggregates, etc. For example, one might
+	// want to monitor not just the requests per second overall, but the
```
per-second

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+	//
+	// This constraint is ignored if the matched request message type
+	// doesn't contain any topic. Maximum size of Topic can be 249
+	// characters as per recent Kafka spec and allowed characters are
```
spec. Allowed characters...

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+	// a-z, A-Z, 0-9, -, . and _
+	// Older Kafka versions had longer topic lengths of 255, but in Kafka 0.10
+	// version the length was changed from 255 to 249. For compatibility
+	// reasons we are using 255
```
255.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+
+	// --------------------------------------------------------------------
+	// Private fields. These fields are used internally and are not exposed
+	// via the
```
via the ... ?

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+
+	// apiKeyInt is the integer representation of expanded Role. It is a
+	// list of all low-level apiKeys to
+	// be expanded as per the value of Role
```
... Role.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+	// be expanded as per the value of Role
+	apiKeyInt KafkaRole
+
+	// apiVersionInt is the integer representation of APIVersion
```
APIVersion.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+// List of Kafka apiKeys which have a topic in their
+// request
```
request.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+)
+
+// List of Kafka apiKey which are not associated with
+// any topic
```
topic.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+	ConsumeRole = "consume"
+)
+
+// KafkaAPIKeyMap is the map of all allowed kafka API keys
```
kafka --> Kafka

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+	"alterconfigs":         33, /* AlterConfigs */
+}
+
+// KafkaReverseApiKeyMap is the map of all allowed kafka API keys
```
kafka --> Kafka

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+// KafkaRole is the list of all low-level apiKeys to
+// be expanded as per the value of Role
```
... Role.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+// be expanded as per the value of Role
+type KafkaRole []int16
+
+// KafkaMaxTopicLen is the maximum character len of a topic.
```
len --> length.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+type KafkaRole []int16
+
+// KafkaMaxTopicLen is the maximum character len of a topic.
+// Older Kafka versions had longer topic lengths of 255, in Kafka 0.10 version
```
... of 255; in Kafka...

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+// KafkaMaxTopicLen is the maximum character len of a topic.
+// Older Kafka versions had longer topic lengths of 255, in Kafka 0.10 version
+// the length was changed from 255 to 249. For compatibility reasons we are
+// using 255
```
255.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+)
+
+// KafkaTopicValidChar is a one-time regex generation of all allowed characters
+// in kafka topic name.
```
kafka --> Kafka.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+var KafkaTopicValidChar = regexp.MustCompile(`^[a-zA-Z0-9\\._\\-]+$`)
+
+// CheckAPIKeyRole checks the apiKey value in the request, and returns true if
+// it is allowed else false
```
returns if it is allowed.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+// GetAPIVersion returns the APIVersion as integer or the bool set to true if
+// any API version is allowed
```
allowed.

ianvernon commented on pkg/policy/api/v3/kafka.go:
```diff
@@ -0,0 +1,293 @@
+	return *kr.apiVersionInt, false
+}
+
+// MapRoleToAPIKey maps the Role to the low level set of APIKeys for that role
```
... low-level ... for that role.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+)
+
+// Rule is a policy rule which must be applied to all endpoints which match the
+// labels contained in the endpointSelector
```
endpointSelector.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+	Description string `json:"description,omitempty"`
+}
+
+type IngressRule struct {
```
Documentation?

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+	FromEntities *EntityRule `json:"fromEntities,omitempty"`
+}
+
+type EgressRule struct {
```
Documentation?

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+	ToServices *ServiceRule `json:"toServices,omitempty"`
+}
+
+type IdentityRule struct {
```
Documentation?

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+	ToPorts  *PortRule `json:"toPorts,omitempty"`
+}
+
+// K8sServiceNamespace is an abstraction for the k8s service + namespace types.
```
k8s --> Kubernetes

I do not think we should not be using abbreviations in publicly-facing documentation.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+
+// K8sServiceNamespace is an abstraction for the k8s service + namespace types.
+type K8sServiceNamespace struct {
+	ServiceName string `json:"serviceName,omitempty"`
```
Documentation?

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+}
+
+type ServiceRule struct {
+	// K8sServiceSelector selects services by k8s labels and namespace
```
Kubernetes labels and namespace.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+type ServiceRule struct {
+	// K8sServiceSelector selects services by k8s labels and namespace
+	K8sServiceSelector *K8sServiceSelectorNamespace `json:"k8sServiceSelector,omitempty"`
+	// K8sService selects service by name and namespace pair
```
... pair.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+	ToPorts    *PortRule            `json:"toPorts,omitempty"`
+}
+
+// PortProtocol specifies an L4 port with an optional transport protocol
```
protocol.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+	// +optional
+	Ports []PortProtocol `json:"anyOf,omitempty"`
+
+	// Rules is a list of additional port level rules which must be met in
```
port-level

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+	Rules *L7Rules `json:"rules,omitempty"`
+}
+
+// L7Rules is a union of port level rule types. Mixing of different port
```
port-level

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+
+// L7Rules is a union of port level rule types. Mixing of different port
+// level rule types is disallowed, so exactly one of the following must be set.
+// If none are specified, then no additional port level rules are applied.
```
port-level

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+
+type PortRules []PortRule
+
+// ServiceSelector is a label selector for k8s services
```
k8s --> Kubernetes

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+// ServiceSelector is a label selector for k8s services
+type ServiceSelector IdentitySelector
+
+// Entities specifies the class of receiver/sender endpoints that do not have individual identities.
```
Fix documentation.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+type Entity string
+
+const (
+	// EntityWorld is an entity that represents traffic external to endpoint's cluster
```
cluster.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+const (
+	// EntityWorld is an entity that represents traffic external to endpoint's cluster
+	EntityWorld Entity = "world"
+	// EntityHost is an entity that represents traffic within endpoint host
```
within an endpoint's host.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+	EntityWorld Entity = "world"
+	// EntityHost is an entity that represents traffic within endpoint host
+	EntityHost Entity = "host"
+	// EntityAll is an entity that represents all traffic
```
all traffic.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+	EntityAll Entity = "all"
+)
+
+// EntitySelectorMapping maps special entity names that come in policies to selectors
```
to selectors.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+// Example: 192.0.2.1/32
+type CIDR string
+
+// CIDRMatchAll is a []CIDR that matches everything
```
everything.

ianvernon commented on pkg/policy/api/v3/rule.go:
```diff
@@ -0,0 +1,235 @@
+	return []CIDR{CIDR("0.0.0.0/0"), CIDR("::/0")}
+}
+
+// L4Proto is a layer 4 protocol name
```
name.

ianvernon commented on pkg/policy/api/v3/rule_validation.go:
```diff
@@ -0,0 +1,325 @@
+}
+
+func (i *IngressRule) sanitize() error {
+	l3Members := map[string]bool{
```
Add a comment specifying what exactly this represents, specifically that we do not support combinations of any of the contents of `L3Members` together.

ianvernon commented on pkg/policy/api/v3/rule_validation.go:
```diff
@@ -0,0 +1,325 @@
+	return nil
+}
+
+// Sanitize sanitizes Kafka rules
```
... rules.

ianvernon commented on pkg/policy/api/v3/selector.go:
```diff
@@ -0,0 +1,215 @@
+	return string(j)
+}
+
+// Hash return hash of the internal json structure that represents the identity selector
```
... returns the hash ... selector.

ianvernon commented on pkg/policy/api/v3/selector.go:
```diff
@@ -0,0 +1,215 @@
+	return false
+}
+
+// NewWildcardIdentitySelector returns a selector that matches on all identities
```
identities.

ianvernon commented on pkg/policy/api/v3/selector.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+// NewESFromK8sLabelSelector returns a new identity selector from the label
+// where it the given srcPrefix will be encoded in the label's keys.
```
... where the given...

ianvernon commented on pkg/policy/api/v3/utils.go:
```diff
@@ -0,0 +1,100 @@
+	return len(rules.HTTP) + len(rules.Kafka)
+}
+
+// Exists returns true if the HTTP rule already exists in the list of rules
```
list of rules.

ianvernon commented on pkg/policy/api/v3/utils.go:
```diff
@@ -0,0 +1,100 @@
+	return false
+}
+
+// Equal returns true if both HTTP rules are equal
```
are equal.

ianvernon commented on pkg/policy/api/v3/utils.go:
```diff
@@ -0,0 +1,100 @@
+	return true
+}
+
+// Exists returns true if the HTTP rule already exists in the list of rules
```
in the list of rules.

ianvernon commented on pkg/policy/api/v3/utils.go:
```diff
@@ -0,0 +1,100 @@
+		k.Topic == o.Topic && k.ClientID == o.ClientID && k.Role == o.Role
+}
+
+// Validate returns an error if the layer 4 protocol is not valid
```
valid.

ianvernon commented on pkg/policy/api/v3/utils.go:
```diff
@@ -0,0 +1,100 @@
+	return nil
+}
+
+// NumRules returns the total number of L7Rules configured in this PortRule
```
PortRule.

ianvernon commented on pkg/policy/api/v3/utils.go:
```diff
@@ -0,0 +1,100 @@
+	return r.Rules.Len()
+}
+
+// ParseL4Proto parses a string as layer 4 protocol
```
protocol.

ianvernon commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,537 @@
+	"github.com/cilium/cilium/pkg/policy/api/v2"
+)
+
+// V2RulesTov3RulesSanitized translates the given v2Rules to the respective v3
```
"to their respective v3 counterparts"

ianvernon commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,537 @@
+}
+
+// V2RulesTov3Rules translates the given v2Rules to the respective v3 Rules.
+// All given v2Rules needs to be sanitized and the returned v3 Rules needs to
```
needs --> need

ianvernon commented on pkg/policy/api/v3/v2tov3.go:
```diff
@@ -0,0 +1,537 @@
+	return v3Rule
+}
+
+func v2ESTov3ES(v2ES *v2.EndpointSelector) *IdentitySelector {
```
I'm not a big fan of these abbreviated function names. I prefer verbosity in this case (ES --> EndpointSelector).

ianvernon commented on pkg/policy/api/v3/v2tov3_test.go:
```diff
@@ -0,0 +1,322 @@
+		args args
+		want *PortRule
+	}{
+	// TODO: Add test cases.
```
I'm not for merging this before we add extensive testing to be sure we translate rules effectively.

ianvernon commented on pkg/k8s/apis/cnp.cilium.io/register.go:
```diff
@@ -0,0 +1,20 @@
+package ciliumio
+
+const (
+	// GroupName is the group name of the cilium network policy
```
cilium network policy --> CiliumNetworkPolicy.

ianvernon commented on pkg/k8s/apis/cnp.cilium.io/utils/utils.go:
```diff
@@ -0,0 +1,220 @@
+)
+
+const (
+	// subsysK8s is the value for logfields.LogSubsys
```
logfields.LogSubsys.

ianvernon reviewed (changes requested):
Changes are mainly documentation related. While they are nitpicky, I think they are necessary to address because this is our publicly facing API and this should be as polished as possible.

I'm going to have to come back and review other parts separately because I found the sheer size of this PR difficult to consume.

aanm committed:
cilium/cmd: allow to read both v2 and v3 policy

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
policy/v*: add LabelSelector nil check

Per the definition of a Rule, EndpointSelector cannot be empty. Ensure this constraint
is met when rules are sanitized.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
api/v2: add Sanitize() to Rules

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/*: refatored v3.EndpointSelector to v3.IdentitySelector

In order to properly detect which policy version is being imported to
cilium we had to change the v3.EndpointSelector to v3.IdentitySelector.

By having an unique, and required, field name on different version, it
will be easier to detect which policy version is being imported.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
api/v3: add initial documentation

Signed-off-by: Ian Vernon <ian@covalent.io>
Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
api/v3: Refator EndpointRequire to IdentityRequirement

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
cnp.cilium.io/v3: add CRD schema validation

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
api/v3: removing CIDR wildcard from L4-only

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
add versioning to the v3 policy

This way we will be able to detect which policy the user is importing as
well it will be clear for the user which policy is being used internally
when the user sees the output of the policy by doing `cilium policy
get`.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s: stop watching for v2 CNP events if the v2 CRD is removed

Signed-off-by: André Martins <andre@cilium.io>
